### PR TITLE
Reject received payments where PaymentRequest Expiry Tag time has been exceeded.

### LINKED ISSUE #748: Eclair accepts payments after payment request timeout period.
**Describe the issue**
I create a payment request and specify a timeout of 1 second.
I check the payment request and it correctly contains the 1 second timeout.
I wait 10 seconds.
I pay the payment request from another node.

**What behaviour did you expect?**
The payment should be rejected.
BOLT#11
_A payee SHOULD NOT accept a payment after timestamp plus expiry._
https://github.com/lightningnetwork/lightning-rfc/blob/master/11-payment-encoding.md#payer--payee-requirements

**What was the actual behaviour (provide screenshots if the issue is GUI-related)?**
 The payment is accepted.

**How reliably can you reproduce the issue, what are the steps to do so?**
100% - steps are above.

**What version of Eclair are you using, where did you get it (website, self-compiled, etc)?**
HEAD as of 26th Oct 2018

**What environment are you observing the error on (OS, JDK version)?**
Linux/1.8

**Any extra information that might be useful in the debugging process.**
This match in LocalPaymentHandler has no logic to check the timeout.
In addition to the check on paymentRequest.amount here:
https://github.com/ACINQ/eclair/blob/2de7c6b07d558b614cabf8c77f77d7aab5f607e4/eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala#L78
We need a check on paymentRequest.expirySeconds in here:
https://github.com/ACINQ/eclair/blob/2de7c6b07d558b614cabf8c77f77d7aab5f607e4/eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala#L88

### PR DESCRIPTION
**Pull request by n1bor**
Fixes #748

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala [hunk 1]
```diff
@@ -85,12 +85,21 @@ class LocalPaymentHandler(nodeParams: NodeParams) extends Actor with ActorLoggin
...
               sender ! CMD_FAIL_HTLC(htlc.id, Right(IncorrectPaymentAmount), commit = true)
             case _ =>
-              log.info(s"received payment for paymentHash=${htlc.paymentHash} amountMsat=${htlc.amountMsat}")
-              // amount is correct or was not specified in the payment request
-              nodeParams.paymentsDb.addPayment(Payment(htlc.paymentHash, htlc.amountMsat, Platform.currentTime / 1000))
-              sender ! CMD_FULFILL_HTLC(htlc.id, paymentPreimage, commit = true)
-              context.system.eventStream.publish(PaymentReceived(MilliSatoshi(htlc.amountMsat), htlc.paymentHash, htlc.channelId))
-              context.become(run(hash2preimage - htlc.paymentHash))
+              paymentRequest.expiry match {
+                case Some(expiry: Long) if (expiry+paymentRequest.timestamp <  Platform.currentTime / 1000) =>
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala]: nitpick: specifying the type of expiry is not needed here, and we'd use space around the `+`

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala [hunk 2]
```diff
@@ -85,12 +85,21 @@ class LocalPaymentHandler(nodeParams: NodeParams) extends Actor with ActorLoggin
...
+                  // we remove expired payment requests every 10 mins. But might still get here if request only just expired.
+                  log.warning(s"rejecting received payment where expiry time in seconds has been exceeded: paymentHash=${htlc.paymentHash} timestamp=${paymentRequest.timestamp}"+
+                    s" expiryTag=$expiry timenow=${Platform.currentTime / 1000}")
+                  // Return unknowPaymentHash so we are deteministic - as if we purged the payment request already we would have sent this.
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala]: typo: deteministic instead of deterministic

Comment (n1bor) [reply eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala]: fixed

**Comment by n1bor** [comment]
Think I get what you are saying.
But then we are collect'ng over all entries in the Map every time we receive a payment rather than every 10mins? Obviously with 5 outstanding payments and 5 payments a minute is fine.

But this scales with square of payments? As if we got 100 times more payments per min we would have 100 times more outstanding and 100 times more iterations over map so 10,000 times more work to do?

My solutions only adds one lookup so scales linearly.

**Comment by n1bor** [comment]
Not sure this is any better/clearer. I think I have done what you suggested.
I just wanted to avoid making it collect over all pending requests for every payment as that would scale very badly if getting 100's payments a minute/second...

**Comment by pm47** [comment]
I took the liberty to commit a new version that addresses your concerns. WDYT?

**Pull request merged**